File: clients/php/examples/test_watch.php
<?php
/**
 * ScrivaDB PHP SDK — Watch streaming example.
 *
 * Subscribes to a Watch stream on a collection, inserts records from a
 * background process (using proc_open / forked child), and prints each
 * change event as it arrives.
 *
 * Start the server first:
 *
 *   make run
 *
 * Then run:
 *
 *   cd clients/php
 *   composer install
 *   php examples/test_watch.php
 *
 * Press Ctrl+C to stop.
 */

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use ScrivaDB\ScrivaDB;

$host   = $argv[1] ?? 'localhost';
$port   = (int)($argv[2] ?? 5433);
$apiKey = $argv[3] ?? 'dev-key';

const WATCH_COLLECTION = 'test_php_watch';

$db = new ScrivaDB($host, $port, $apiKey);

// ── Setup ────────────────────────────────────────────────────────────────────
$existing = $db->listCollections();
if (in_array(WATCH_COLLECTION, $existing, true)) {
    $db->dropCollection(WATCH_COLLECTION);
}
$db->createCollection(WATCH_COLLECTION);
echo "Watching collection '" . WATCH_COLLECTION . "' — inserting 3 records from a child process ...\n\n";

// ── Fork a child to do inserts after a short delay ───────────────────────────
$pid = pcntl_fork();
if ($pid === -1) {
    // pcntl_fork not available — fall back to inserting in the same process
    // after starting the watch (this won't actually work for long-running watch,
    // but demonstrates the API shape).
    $pid = 0;
}

if ($pid === 0 && function_exists('pcntl_fork')) {
    // Child process: sleep briefly then insert records
    usleep(300_000); // 300ms

    $child = new ScrivaDB($host, $port, $apiKey);
    $child->insert(WATCH_COLLECTION, ['event' => 'one',   'ts' => time()]);
    usleep(200_000);
    $child->insert(WATCH_COLLECTION, ['event' => 'two',   'ts' => time()]);
    usleep(200_000);
    $child->update(WATCH_COLLECTION, 1, ['event' => 'one-updated', 'ts' => time()]);
    usleep(200_000);
    $child->delete(WATCH_COLLECTION, 2);
    exit(0);
}

// ── Parent: consume Watch stream (stops after 4 events) ──────────────────────
$maxEvents = 4;
$received  = 0;

foreach ($db->watch(WATCH_COLLECTION) as $event) {
    echo sprintf(
        "[%s] op=%-10s id=%s data=%s\n",
        $event['ts'] ?? 'n/a',
        $event['op'],
        $event['record']['id'] ?? '?',
        json_encode($event['record']['data'] ?? [])
    );
    ++$received;
    if ($received >= $maxEvents) {
        break;
    }
}

// ── Cleanup ───────────────────────────────────────────────────────────────────
if ($pid > 0 && function_exists('pcntl_waitpid')) {
    pcntl_waitpid($pid, $childStatus);
}

$db->dropCollection(WATCH_COLLECTION);
echo "\nWatch example complete. Received $received events.\n";

File: clients/rust/examples/test_watch.rs
//! test_watch — watch example for the ScrivaDB Rust client.
//!
//! Subscribes to change events on a collection while inserting records in a
//! background task, then prints each event received.
//!
//! Prerequisites:
//!   - ScrivaDB server running: `make run` from the repo root.
//!   - Run: `cargo run --example test_watch` from clients/rust/.

use futures::StreamExt;
use scriva::{ScrivaDB, FilterInput, FilterOp};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    // Ensure the collection exists before watching.
    let mut setup = ScrivaDB::connect("localhost", 5433, "dev-key").await?;
    // Drop from any previous run.
    let _ = setup.drop_collection("watch_rs").await;
    setup.create_collection("watch_rs").await?;
    drop(setup);

    // Spawn a background task that inserts records after a short delay.
    let insert_handle = tokio::spawn(async {
        tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
        let mut db = ScrivaDB::connect("localhost", 5433, "dev-key").await.unwrap();
        db.insert("watch_rs", serde_json::json!({"name": "Alice", "role": "admin"})).await.unwrap();
        db.insert("watch_rs", serde_json::json!({"name": "Bob",   "role": "user"})).await.unwrap();
        db.insert("watch_rs", serde_json::json!({"name": "Carol", "role": "admin"})).await.unwrap();

        let id = db.insert("watch_rs", serde_json::json!({"name": "Dave"})).await.unwrap();
        db.update("watch_rs", id, serde_json::json!({"name": "Dave", "role": "user"})).await.unwrap();
        db.delete("watch_rs", id).await.unwrap();

        println!("[inserter] done — closing connection");
    });

    // Watch only admin events using a filter.
    let mut watcher = ScrivaDB::connect("localhost", 5433, "dev-key").await?;
    let filter = FilterInput::field("role", FilterOp::Eq, "admin");
    let mut stream = watcher.watch("watch_rs", Some(filter)).await?;

    println!("[watcher] subscribed to watch_rs (role=admin only)");
    println!("[watcher] waiting for events — press Ctrl-C to stop");

    let mut count = 0usize;
    loop {
        tokio::select! {
            maybe_event = stream.next() => {
                match maybe_event {
                    Some(Ok(event)) => {
                        println!(
                            "[watcher] {:?} collection={} id={} data={}{}",
                            event.op,
                            event.collection,
                            event.record.id,
                            event.record.data,
                            event.ts.as_deref().map(|t| format!(" ts={}", t)).unwrap_or_default(),
                        );
                        count += 1;
                        if count >= 2 {
                            // Received both admin inserts — exit gracefully.
                            break;
                        }
                    }
                    Some(Err(status)) => {
                        eprintln!("[watcher] error: {}", status);
                        break;
                    }
                    None => {
                        println!("[watcher] stream ended");
                        break;
                    }
                }
            }
            _ = tokio::time::sleep(tokio::time::Duration::from_secs(10)) => {
                println!("[watcher] timed out waiting for events");
                break;
            }
        }
    }

    insert_handle.await?;

    // Cleanup
    let mut cleanup = ScrivaDB::connect("localhost", 5433, "dev-key").await?;
    cleanup.drop_collection("watch_rs").await?;
    println!("[watcher] collection dropped, exiting");
    Ok(())
}
